cmd/estuary-shuttle/main.go

```diff
@@ -91,154 +158,156 @@ func main() {
-			Required: true,
+			Name:  "auth-token",
+			Usage: "auth token for connecting to estuary",
+			Value: cfg.Estuary.AuthToken,
 		},
 		&cli.StringFlag{
-			Name:     "handle",
-			Usage:    "estuary shuttle handle to use",
-			Required: true,
+			Name:  "handle",
+			Usage: "estuary shuttle handle to use",
+			Value: cfg.Estuary.Handle,
 		},
 		&cli.StringFlag{
 			Name:  "host",
 			Usage: "url that this node is publicly dialable at",
+			Value: cfg.Hostname,
 		},
 		&cli.BoolFlag{
-			Name: "logging",
+			Name:  "logging",
+			Value: cfg.Logging.ApiEndpointLogging,
 		},
 		&cli.BoolFlag{
-			Name: "write-log-flush",
+			Name:  "write-log-flush",
+			Value: cfg.Node.HardFlushWriteLog,
 		},
 		&cli.BoolFlag{
-			Name: "write-log-truncate",
+			Name:  "write-log-truncate",
+			Value: cfg.Node.WriteLogTruncate,
 		},
 		&cli.BoolFlag{
-			Name: "no-blockstore-cache",
+			Name:  "no-blockstore-cache",
+			Value: cfg.Node.NoBlockstoreCache,
 		},
 		&cli.BoolFlag{
-			Name: "private",
+			Name:  "private",
+			Value: cfg.Private,
 		},
 		&cli.BoolFlag{
 			Name:  "disable-local-content-adding",
 			Usage: "disallow new content ingestion on this node",
+			Value: cfg.Content.Disable,
 		},
 		&cli.BoolFlag{
-			Name: "no-reload-pin-queue",
+			Name:  "no-reload-pin-queue",
+			Value: cfg.NoReloadPinQueue,
 		},
 		&cli.BoolFlag{
 			Name:  "dev",
 			Usage: "use http:// and ws:// when connecting to estuary in a development environment",
+			Value: cfg.Dev,
 		},
 		&cli.StringSliceFlag{
 			Name:  "announce-addr",
 			Usage: "specify multiaddrs that this node can be connected to on",
+			Value: cli.NewStringSlice(cfg.Node.AnnounceAddrs...),
 		},
 		&cli.BoolFlag{
 			Name:  "jaeger-tracing",
-			Value: false,
+			Value: cfg.Jaeger.JaegerTracing,
 		},
 		&cli.StringFlag{
 			Name:  "jaeger-provider-url",
-			Value: "http://localhost:14268/api/traces",
+			Value: cfg.Jaeger.JaegerProviderUrl,
 		},
 		&cli.Float64Flag{
 			Name:  "jaeger-sampler-ratio",
 			Usage: "If less than 1 probabilistic metrics will be used.",
-			Value: 1,
+			Value: cfg.Jaeger.JaegerSamplerRatio,
 		},
 		&cli.BoolFlag{
 			Name:  "libp2p-websockets",
-			Value: false,
+			Value: cfg.Node.HasListener(config.DefaultWebsocketAddr),
 		},
 		&cli.Int64Flag{
 			Name:  "bitswap-max-work-per-peer",
-			Value: 5 << 20,
+			Value: cfg.Node.BitswapConfig.MaxOutstandingBytesPerPeer,
 		},
 		&cli.IntFlag{
 			Name:  "bitswap-target-message-size",
-			Value: 16 << 10,
+			Value: cfg.Node.BitswapConfig.TargetMessageSize,
 		},
 	}
 
-	app.Action = func(cctx *cli.Context) error {
-		ddir := cctx.String("datadir")
-
-		bsdir := cctx.String("blockstore")
-		if bsdir == "" {
-			bsdir = filepath.Join(ddir, "blocks")
-		} else if bsdir[0] != '/' && bsdir[0] != ':' {
-			bsdir = filepath.Join(ddir, bsdir)
-		}
-
-		wlog := cctx.String("write-log")
-		if wlog != "" && wlog[0] != '/' {
-			wlog = filepath.Join(ddir, wlog)
-		}
+	app.Commands = []*cli.Command{
```

Comment:
I think the approach below is more intuitive, and we can easily expand the `config` command to have more sub-commands in the future. Also, I think the code should own the config filename and users should only specify the path.

```
app.Commands = []*cli.Command{
	{
		Name: "config",
		Subcommands: []*cli.Command{
			{
				Name:  "init",
				Usage: "creates a configuration file in the location specified by the path parameter",
				Flags: []cli.Flag{
					&cli.StringFlag{
						Name:     "path",
						Usage:    "specify path for where the config file with be created",
						Required: true,
					},
				},
				Action: func(cctx *cli.Context) error {
					path := cctx.String("path")
					if err := cfg.LoadFromPath(path, true); err != nil {
						return err
					}
					return overrideConfigWithDefaults(app.Flags, cctx, cfg) // call cfg.Save(configuration) in overrideWithDefaults - given we will take this line out in the future
				},
			},
		},
	},
}
```

This way, we can have `./estuary-shuttle  config init --path=` to create the config file `estuary-shuttle.json` in `path`

Reply:
I feel this verges on extra functionality. I was trying to get existing flags working in a consistent way in this PR. 

Agree that config is in the long term likely to have a bunch of subcommands. Before launching into that though I think we should maybe discuss how the various pieces of estuary might interact (for example, to streamline the process of registering a shuttle).

Reply:
@softwareplumber I still think the code should own the config file name, and users should specify only dir path.